File: cse216hw4/BinaryTree.java
import java.util.Iterator;
import java.util.LinkedList;
import java.util.NoSuchElementException;
import java.util.Queue;

public class BinaryTree<T> implements Iterable<T> {

    T data;
    BinaryTree<T> leftChild;
    BinaryTree<T> rightChild;

    public BinaryTree() {
        this.data = null;
    }

    public BinaryTree(T data) {
        this.data = data;
    }

    public void addLeftChild(BinaryTree childTree) {
        this.leftChild = childTree;
    }

    public void addRightChild(BinaryTree childTree) {
        this.rightChild = childTree;
    }

    @Override
    public Iterator<T> iterator() {
        return new BinaryTreeIterator(this);
    }

    private class BinaryTreeIterator implements Iterator<T> {

        private Queue<BinaryTree<T>> queue = new LinkedList<>();

        public BinaryTreeIterator(BinaryTree<T> tree) {
            if (tree.data != null)
                this.queue.add(tree);
        }
        
        @Override
        public boolean hasNext() {
            return !this.queue.isEmpty();
        }

        @Override
        public T next() {
            if (!this.hasNext()) {
                throw new NoSuchElementException();
            }
            BinaryTree<T> tree = queue.remove();
            if (tree.leftChild != null) {
                queue.add(tree.leftChild);
            }
            if (tree.rightChild != null) {
                queue.add(tree.rightChild);
            }
            return tree.data;
        }
    }
}

File: cse216hw4/binarytree.py
class BinaryTree:
    def __init__(self, data=None):
        self.data = data
        self.left = None
        self.right = None

    @property
    def data(self):
        return self.__data

    @data.setter
    def data(self, data):
        self.__data = data

    def __iter__(self):
        if self.data is not None:
            yield self.data
        if self.left is not None:
            for child in self.left.__iter__():
                yield child
        if self.right is not None:
            for child in self.right.__iter__():
                yield child

    def add_leftchild(self, tree):
        if type(self.data) != type(tree.data):
            raise TypeError("Type mismatch between "+type(self.data).__name__+" and "+type(tree.data).__name__)
        else:
            self.left = tree

    def add_rightchild(self, tree):
        if type(self.data) != type(tree.data):
            raise TypeError("Type mismatch between "+type(self.data).__name__+" and "+type(tree.data).__name__)
        else:
            self.right = tree